docs: attempt to clarify magic methods supported by PyO3 (#4190)
PyO3/pyo3: guide/src/class/protocols.md

If a magic method is not on the list above (for example `__init_subclass__`), then it should just work in PyO3. If this is not the case, please file a bug report.

## Magic Methods handled by PyO3

If a function name in `#[pymethods]` is a magic method which is known to need special handling, it will be automatically placed into the correct slot in the Python type object. The function name is taken from the usual rules for naming `#[pymethods]`: the `#[pyo3(name = "...")]` attribute is used if present, otherwise the Rust function name is used.

The magic methods handled by PyO3 are very similar to the standard Python ones on [this page](https://docs.python.org/3/reference/datamodel.html#special-method-names) - in particular they are the subset which have slots as [defined here](https://docs.python.org/3/c-api/typeobj.html).

When PyO3 handles a magic method, a couple of changes apply compared to other `#[pymethods]`:
 - The Rust function signature is restricted to match the magic method.
 - The `#[pyo3(signature = (...)]` and `#[pyo3(text_signature = "...")]` attributes are not allowed.

> The following sections list of all magic methods PyO3 needs to handle.  The

Reviewer: I feel like this does not really get over the message we are trying to make (that we implement the special handling).

```suggestion
The following sections list all magic methods for which PyO3 implements the necessary special handling.  The
```

Author: Agreed, your phrasing is much clearer 👍
